### PR TITLE
feat: ✨ add flag to disable getSession after signIn on local / refresh provider

### PR DESCRIPTION
**Pull request by bitfactory-frank-spee**
### 🔗 Linked issue

https://github.com/sidebase/nuxt-auth/issues/701

### ❓ Type of change

- [x] 📖 Documentation (updates to the documentation, readme or JSdoc annotations)
- [x] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

See issue for clarification. This adds a flag to the signInOptions named `withSession` which defaults to true. This makes sure this will be backwards compatible. When `withSession === false` the `getSession` method will not be called.

This also includes a bug fix (681fdb07218397fbe52bf097b6daf044762970a8) where refresh was not equal to local provider.

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [ ] I have added tests (if possible).
- [x] I have updated the documentation accordingly.

NOTE: this documentation mentioned `withGetSession` before and not `withSession` as decided to be the name later.

### CODE REVIEW

**Review by bitfactory-oussama-fadlaoui — APPROVED** [review]

**Comment by phoenix-ru** [comment]
Hi @bitfactory-frank-spee , thanks for your contribution, it looks good to me.

I will come back to it after #694 , so that we can ensure that it's functionally good.

**Comment by phoenix-ru** [comment]
@bitfactory-frank-spee Could you please sync your PR with the current main?

**Comment by phoenix-ru** [comment]
bump @bitfactory-frank-spee

**Comment by bitfactory-frank-spee** [comment]
@phoenix-ru I will update it this week 👍

**Comment by bitfactory-frank-spee** [comment]
> @bitfactory-frank-spee Could you please sync your PR with the current main?

Done 🚀

File: docs/content/3.application-side/2.session-access-and-management.md [hunk 1]
```diff
@@ -103,12 +105,16 @@ await signIn(credentials, { callbackUrl: '/protected' })
 // Trigger a sign-in with a redirect afterwards to an external page (if set, this will cause a hard refresh of the page)
 await signIn(credentials, { callbackUrl: 'https://sidebase.io', external: true })
 
+// Trigger a sign-in without calling getSession directly. You have to manually call it to get session data.
+await signIn(credentials, { withGetSession: false })
```
Comment (zoey-kaiser) [inline docs/content/3.application-side/2.session-access-and-management.md:109]: Can we rename this new option from `withGetSession` to just `getSession`? I think this would be a bit more readable!

**Review by zoey-kaiser — CHANGES_REQUESTED** [review]
Hi @bitfactory-frank-spee 👋 

Very close to being done! I have one minor renaming request and then we should be good to go, to release this in 0.8!

File: docs/content/3.application-side/2.session-access-and-management.md (hunk 1 above)
Comment (bitfactory-frank-spee) [reply docs/content/3.application-side/2.session-access-and-management.md:109]: Hi Zoey, you just caught me on holiday 🌴. I will change it asap, next week for certain. I will try to do it sooner. 

Renaming is fine by me. I don't totally agree with the name change. Because it is a boolean you want to name it like `is...`, `has...`, `with...` or something like that, and not a get/set/action like method name. 

So I would suggest just `withSession` or without 'with', e.g. something like `callSession` or just `session`. But again all names are fine by me. Just giving my opinion to get it as good as possible.

Comment (zoey-kaiser) [reply docs/content/3.application-side/2.session-access-and-management.md:109]: Agreed! Let's go with `withSession`

Comment (bitfactory-frank-spee) [reply docs/content/3.application-side/2.session-access-and-management.md]: Done in [`ac9415f` (#702)](https://github.com/sidebase/nuxt-auth/pull/702/commits/ac9415fb6ff552478872c391badbdd1d5bf11d86)

**Comment by zoey-kaiser** [comment]
Hi @bitfactory-frank-spee 👋 

We just released 0.8.0, so I would begin looking into getting this PR merged in the next release. Could you please merge main and we can then do a final round of reviews 😊

File: src/runtime/composables/local/useAuth.ts [hunk 2]
```diff
@@ -32,9 +32,12 @@ const signIn: SignInFunc<Credentials, any> = async (credentials, signInOptions,
   const { rawToken } = useAuthState()
   rawToken.value = extractedToken
 
-  await nextTick(getSession)
+  const { callbackUrl, redirect = true, external, withSession = true } = signInOptions ?? {}
+
+  if (withSession) {
```
Comment (zoey-kaiser) [inline src/runtime/composables/local/useAuth.ts:37]: Could you please add type support for the new `withSession` option? Please also consider, that you will need to account for the fact that this value is only available for the `local` and `refresh` providers and should therefore not be added to the signIn function of the `authjs` provider.

**Review by zoey-kaiser — CHANGES_REQUESTED** [review]
Just some typings missing 😊

**Comment by Mucaccino** [comment]
I need this. Really this was gone stale?

File: src/runtime/composables/local/useAuth.ts (hunk 2 above)
Comment (bitfactory-frank-spee) [reply src/runtime/composables/local/useAuth.ts:37]: Hi Zoey. Sorry it took a while. In my mind I responded somewhere. I can add typing.

But can you direct / help me to how I do this "account for the fact that this value is only available for the local and refresh providers and should therefore not be added to the signIn function of the authjs provider"? Because I have no idea what you mean here.

**Comment by bitfactory-frank-spee** [comment]
> I need this. Really this was gone stale?

Ahh, I now see that I needed to respond to this, and that I didn't. I have now: https://github.com/sidebase/nuxt-auth/pull/702#discussion_r1810021411

**Comment by bitfactory-frank-spee** [comment]
I merged the current main into this PR and fixed the conflicts as the local and refresh provider are merged in the new release. I also fixed the documentation.

File: src/runtime/composables/local/useAuth.ts (hunk 2 above)
Comment (bitfactory-frank-spee) [reply src/runtime/composables/local/useAuth.ts:60]: So where do I add the typing support?

Comment (bitfactory-frank-spee) [reply src/runtime/composables/local/useAuth.ts:60]: And how do I make sure this is not added to the signIn function of the authjs provider?

Comment (bitfactory-frank-spee) [reply src/runtime/composables/local/useAuth.ts:60]: Is this enough? https://github.com/sidebase/nuxt-auth/pull/702/files#diff-23ace85c1398da87fad352ed08fd7026dab0a96402ccc8b6c3e23b478c993f43

File: src/runtime/types.ts [hunk 3]
```diff
@@ -548,6 +548,12 @@ export interface SecondarySignInOptions extends Record<string, unknown> {
...
   external?: boolean
+  /**
+   * Whether `getSession` is called. If set to false, you have to manually call `getSession` to get the session data.
+   *
+   * @default true
+   */
+  withSession?: boolean
 }
 
 export interface SignUpOptions extends SecondarySignInOptions {
```
Comment (bitfactory-frank-spee) [inline src/runtime/types.ts:559]: This interface is odd IMHO. Because I never see this used in the code with any of the properties of the `SecondarySignInOptions` interface. It is just used next to a property which uses `SecondarySignInOptions`. But I could miss something which is there somewhere.

File: src/runtime/composables/local/useAuth.ts (hunk 2 above)
Comment (bitfactory-frank-spee) [reply src/runtime/composables/local/useAuth.ts:60]: @zoey-kaiser ?

File: docs/guide/application-side/session-access.md [hunk 4]
```diff
@@ -102,7 +102,7 @@ Time at which the session was last refreshed, either `undefined` if no refresh w
 
 Returns the current Cross Site Request Forgery Token (CSRF Token) required to make POST requests (e.g. for signing in and signing out).
 
-You likely only need to use this if you are not using the built-in `signIn()` and `signOut()` methods. Read more: https://next-auth.js.org/getting-started/client#getcsrftoken
+You likely only need to use this if you are not using the built-in `signIn()` and `signOut()` methods. Read more: <https://next-auth.js.org/getting-started/client#getcsrftoken>
```
Comment (phoenix-ru) [inline docs/guide/application-side/session-access.md:105]: Why did you change the link format? This is a markdown file, and links are correctly recognized by Vitepress
```suggestion
You likely only need to use this if you are not using the built-in `signIn()` and `signOut()` methods. Read more: https://next-auth.js.org/getting-started/client#getcsrftoken
```

**Review by phoenix-ru — COMMENTED** [review]
Hi @bitfactory-frank-spee , sorry for us not answering for a long time, I would take over the PR and make sure it's over the line :)

**Comment by phoenix-ru** [comment]
@bitfactory-frank-spee, unfortunately, I cannot directly commit to your PR (probably due to restrictions on your org). I would have to create a branch off your PR to proceed

File: src/runtime/types.ts (hunk 3 above)
Comment (phoenix-ru) [reply src/runtime/types.ts:559]: You're actually right, it's a weird implementation in https://github.com/sidebase/nuxt-auth/pull/615 which accepts two identical interfaces:
https://github.com/sidebase/nuxt-auth/blob/218a8465fd01fbf70562ab4cc1083dea3931d1a1/src/runtime/composables/local/useAuth.ts#L161

It should be just `signUpOptions`, will probably move this to a different breaking change

Comment (phoenix-ru) [reply src/runtime/types.ts:559]: Issue created: https://github.com/sidebase/nuxt-auth/issues/948

File: src/runtime/types.ts [hunk 5]
```diff
@@ -548,6 +548,12 @@ export interface SecondarySignInOptions extends Record<string, unknown> {
    * @default false
    */
   external?: boolean
+  /**
+   * Whether `getSession` is called. If set to false, you have to manually call `getSession` to get the session data.
+   *
+   * @default true
+   */
```
Comment (phoenix-ru) [inline src/runtime/types.ts:555]:
```suggestion
  /**
   * Whether `getSession` needs to be called after a successful sign-in. When set to false, you can manually call `getSession` to obtain the session data.
   *
   * @default true
   */
```

File: src/runtime/types.ts [hunk 6]
```diff
@@ -548,6 +548,12 @@ export interface SecondarySignInOptions extends Record<string, unknown> {
...
+  withSession?: boolean
```
Comment (phoenix-ru) [inline src/runtime/types.ts:556]: I am actually not in favor of this name, as I used it when testing your PR and it doesn't "click". Existing options use the imperative forms and I think we should stick to it for the sake of consistency (e.g. `redirect`, `preventLoginFlow`).

Another problem I see with `withSession` is that it may confuse users as to what it means. I propose a more explicit

```suggestion
  callGetSession?: boolean
```

**Comment by bitfactory-frank-spee** [comment]
> Hi @bitfactory-frank-spee , sorry for us not answering for a long time, I would take over the PR and make sure it's over the line :)

No problem. I will check your comments today.

File: src/runtime/types.ts (hunk 3 above)
Comment (bitfactory-frank-spee) [reply src/runtime/types.ts:559]: Excellent thanks 👍

**Comment by bitfactory-frank-spee** [comment]
> @bitfactory-frank-spee, unfortunately, I cannot directly commit to your PR (probably due to restrictions on your org).
> 
> I still checked it though and it seems to work, but I have several points

Ah that could be. I am sorry about that. I can probably not change that immediately. Next time I will use my personal account.

File: src/runtime/types.ts (hunk 6 above)
Comment (bitfactory-frank-spee) [reply src/runtime/types.ts:556]: There was an earlier thread with Zoey about this. We ended up with `withSession`. I think it's fine to change, I will do it. See https://github.com/sidebase/nuxt-auth/pull/702/files/e59026bf2b8492e382b04ba8f1f7e12bf072f601#diff-dd4f118f099f41a809eb53d231259e96a9818b3d7f734b4b671de323ac718685

File: docs/guide/application-side/session-access.md (hunk 4 above)
Comment (bitfactory-frank-spee) [reply docs/guide/application-side/session-access.md:105]: Ahh, that is probably markdownlint. I highly recommend to implement this. But I will remove them for now.

Comment (bitfactory-frank-spee) [reply docs/guide/application-side/session-access.md:105]: It is this rule: https://github.com/DavidAnson/markdownlint/blob/v0.36.1/doc/md034.md

Comment (bitfactory-frank-spee) [reply docs/guide/application-side/session-access.md]: https://github.com/sidebase/nuxt-auth/pull/702/commits/7739fbc1d06753c7f69649aab4011347fa6a43a2

**Comment by bitfactory-frank-spee** [comment]
@phoenix-ru I fixed everything you asked. And updated this branch with the main branch.

**Review by phoenix-ru — APPROVED** [review]

**Comment by bitfactory-frank-spee** [comment]
> Hi @bitfactory-frank-spee , extremely sorry for deprioritizing this for a long time! I will ask @zoey-kaiser if she's fine with squeezing this PR into the upcoming `0.10.0` release and merge asap

Excellent! 🎉 🚀 Thanks for the help. Good luck with further releases.
